File: backend/models/portfolio.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

// Type enum of portfolio items
type Type string

const (
	File   Type = "file"
	Folder Type = "folder"
)

// Type language for portfolio items (of supported languages for frontend rendering)
type Language string

const (
	Markdown Language = "markdown"
	Python   Language = "python"
	Ruby     Language = "ruby"
	Ini      Language = "ini"
	Json     Language = "json"
	Pdf      Language = "pdf"
	Default  Language = "default"
)

type FileTreeEntry struct {
	ID       uuid.UUID `gorm:"type:uuid;primary_key;" json:"id"`
	Name     string    `gorm:"not null" json:"name"`
	Type     Type      `gorm:"not null" json:"type"`
	Language Language  `gorm:"type:text;not null" json:"language"`
	Content  string    `gorm:"type:text" json:"content"`
	ParentID uuid.UUID `gorm:"type:uuid;index" json:"-"`

	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"-"`

	Parent   *FileTreeEntry   `gorm:"foreignKey:ParentID" json:"-"`
	Children []*FileTreeEntry `gorm:"foreignKey:ParentID" json:"children"`
}

func (entry *FileTreeEntry) BeforeCreate(tx *gorm.DB) error {
	if entry.ID == uuid.Nil {
		entry.ID = uuid.New()
	}
	return nil
}

File: frontend/types/portfolio.ts
import { ElementType } from 'react';

// Defines the structure for a single file or folder entry in the sidebar tree.
export interface FileTreeEntry {
  id: string; // Unique identifier (e.g., 'docs-readme')
  name: string; // Display name (e.g., 'README.md')
  type: 'file' | 'folder';
  icon: ElementType; // Lucide icon component (passed as ElementType)
  language: 'markdown' | 'python' | 'ruby' | 'ini' | 'pdf' | 'json' | 'default'; // The rendering format
  content: string; // The raw text, code, or URL content
  children?: FileTreeEntry[]; // Required only if type is 'folder'
}

// Defines the valid states for the main activity bar menu
export type ActiveMenuType = 'explorer' | 'search' | 'settings';

// Defines the props for the reusable VSCodeIcon component
export interface VSCodeIconProps {
  icon: ElementType;
  label: string;
  isActive?: boolean;
  onClick: () => void;
}